Documentation/design-docs/tailcalls-with-helpers.md

```diff
@@ -0,0 +1,266 @@
+* TailCallHelperStack - a per thread stack of helper entries that is used to determine whether a tail call is chained or not. Its entries are allocated as local variables in CallTarget thunks. Each entry contains:
+  * Stack pointer captured right before a call to a tail call target
+  * ChainCall flag indicating whether the CallTarget thunk should return after the call to the tail call target or whether it should execute its epilog and jump to TailCallHelper instead. The latter is used by the TailCallHelper to remove the stack frame of the CallTarget before making a tail call from a tail called function. 
+  * Pointer to the next entry on the stack.
+
+### Tail call sequence
+* The caller calls the StoreArguments thunk corresponding to the callee to store the arguments, their GC descriptors and the corresponding CallTarget thunk address in a thread local storage.
+* The caller executes its epilog, restoring stack pointer and callee saved registers to their values when the caller was entered.
+* The caller jumps to the TailCallHelper. This function performs the following operations:
+  * Get the topmost TailCallHelperStack entry for the current thread.
+  * Check if the previous stack frame is a CallTarget thunk frame by comparing the stack pointer value stored in the TailCallHelperStack entry to the current CFA (call frame address). If it matches, it means that the previous stack frame belongs to a CallTarget thunk and so the tail call caller was also tail called.
+  * If the previous frame was a CallTarget thunk, its stack frame needs to be removed to ensure that the stack will not grow when tail calls are chained. Set ChainCall flag in the TailCallHelperStack entry and return. That returns control to the CallTarget thunk. It checks the ChainCall flag and since it set, it executes its epilog and jumps to the TailCallHelper again.
+  * If the previous frame was not a CallTarget thunk, get the address of the CallTarget thunk of the tailcall target from the arguments buffer and jump to it.
+* The CallTarget thunk function then does the following operation:
+  * Create local instance of TailCallHelperStack entry and store the current stack pointer value in it.
+  * Push the entry to the TailCallHelperStack of the current thread.
+  * Get the arguments buffer from the thread local storage, extract the arguments, release the buffer and call the target function. The frame of the CallTarget thunk ensures that the arguments of the target are GC protected until the target function returns or tail calls to another function.
+  * Pop the TailCallHelperStack entry from the TailCallHelperStack of the current thread.
+  * Check the ChainCall flag in the TailCallHelperStack entry. If it is set, run epilog and jump to the TailCallHelper.
+  * If the ChainCall flag is clear, it means that the last function in the tail call chain has returned. So return the return value of the target function.
```

Comment:
I am still trying to understand how struct return via retBuf arg are handled in the new tail call mechanism.  I am writing down my understanding below if we were to implement second idea from @jkotas.

say A() -> call B() -> tail calls C()
where B() and C() return a value type passed via hidden RetBuf arg.

A() would allocate a struct local for return value and pass a byref to it to B().
B() would first call StoreArgument() IL thunk with all the arguments as if it were calling C() except that it considers its return value to be void.  

If B() were to call C() normally, jitted code of B() is free to allocate another local to receive return value from C and copy that back to RetBuf that A() has passed to B().  In case of B() tail calling C(), we want B() to pass Retbuf arg that A() passed to it to StoreArgument() thunk.  Note C() might require other hidden args, example generic context those too needs to be passed to StoreArgument() thunk.  Vararg cookie may not be interesting since we don't support tail calling vararg methods.  There could be other non-standard hidden args (e.g. Virtual Stub Dispatch (VSD) param R11 on Amd64).

Side Question: How do we see that CallTarget IL thunk passing other hidden args like generics cookie to C(), given CallTarget IL thunk is JIT compiled like any other managed method?  Also , we need layout how the new proposal works tail calling VSD calls that require hidden arg passed in R11.

StoreArgument() thunk stores args in TLS and branches to TailCallHelper, which in turn copies args from TLS to stack and branches to CallTarget() IL thunk.  Given CallTarget() is an IL thunk jit compiled like any other managed method, while calling C(), it would allocate a local for receiving retVal from C() and copy that to the RetBuf passed by B() in args.

A() -> CallTarget() of C -> C()

In case of single level tail call C() would return struct retval by copying into RetBuf that was passed by A() to B() and it would work correctly.  

Say C() in turn tail calls into D.
C() --> StoreArgument() thunk - C() would have to pass RetBuf arg that in turn received from its caller to D()
C() --> branches to TailCallHelper
TailCallHelper --> marks as chained tail call and returns to CallTarget() of C by executing its epilog.
Before returning TailCallHelper is required to zero init at least gc-ref fields of RetBuf arg
CallTarget() of C --> copy return value to RetBuf that it received from B() and branches to TailCallHelper after executing its epilog.
TailCallHelper --> branches to CallTarget of D
CallTarget of D --> calls D() and copies the return value it received from D() into RetBuf arg that it received from C() which is same as RetBuf arg C() received from B() which is same as RetBuf arg it received from A().
Finally when CallTarget of D returns to A(), A() would find the return value in RetBuf arg that it passed and things work correctly.

Regarding @janvorli's idea of having TailCallHelper as having the same return type as the target of the tail call:
I thought TailCallHelper is an asm helper that is common to all tail call targets. This would mean now we will have one TailCallHelper per tail call target, right?  

Revisiting above example of A() -> call B() -> tail calls C().
B() -> StoreArgument() thunk - so B() would not pass RetBuf arg in arguments to the thunk (is that correct?)
B() -> branches to TailCallHelper by passing it RetBuf arg that it received from A()
TailCallHelper() -> setup args for CallTarget of C() on stack and branches to it
CallTarget of C() -> calls C() passing all required args

Question: Please help me understand what advantages we see having TailCallHelper having the same return type as target of the tail call over passing RetBuf to StoreArguments Thunk?